Implement Data table with search field and pagination

titkov49 opened the pull request.

Pe5h4 commented on demo/src/modules/home/containers/TableContainer.js:
```diff
@@ -0,0 +1,218 @@
+import { Container } from 'reactstrap';
+
+export default function (props) {
+  const headers = [ { name: 'name', type: 'string' }, { name: 'provider', type: 'string' }, { name: 'type', type: 'string' }, { name: ' ', type: 'Json' }, ];
```
@titkov49 Please, use tabs instead of spaces in the containers:
demo/src/modules/home/containers/TableContainer.js
src/containers/DataTable/Pagination.js
src/containers/DataTable/index.js 

And please add a newline at the end of the files

Pe5h4 commented on doc/datatable.md:
```diff
@@ -0,0 +1,66 @@
+# ASAB WebUI DataTable component
```
@titkov49 I miss the information, how one can add a datatable component to an existing project. Also there is no mention here about that it needs to be added to navigation (routes).

It would be great, if you will also mention, that the developer needs to implement axios and define api call(s) to get the data. In demo, you have solved it with mocked ones, but this will not be the case when used in the products.

Also some example of api call with the use of pagination would be great. Now you are paginating through the mocked data, which are static, but this will not be the case in 99% of the usage of this data table component.

Pe5h4 commented on demo/src/modules/home/containers/TableContainer.js:
```diff
@@ -2,217 +2,167 @@ import React, { useState, useEffect } from 'react';
-  )
-};
\ No newline at end of file
+const mapStateToProps = state => ({ config:  state.config })
```
@titkov49 Pleas add a new line at the end of the file

Pe5h4 commented on doc/datatable.md:
```diff
@@ -1,27 +1,36 @@
 
 # Implementation
 
+DataTable can be mounted in its own container as it is implemented in demo app or as a widget in another app's container. In first case you must create container for `DataTable` and connect it to Home Module of your application by adding it's route and navigation item. Example may be found in `demo/src/modules/home/index.js`.
```
@titkov49 It does not need to be connected to HomeModule of the application, it can be connected to **any** module. Also I would put there an example how the DataTable component is implemented into the project, eg.:

```
import HomeContainer from './containers/HomeContainer'
import Module from 'asab-webui/abc/Module';
import TableContainer from './containers/TableContainer';


export default class HomeModule extends Module {
	constructor(app, name){
		super(app, "HomeModule");
		app.Router.addRoute({ path: '/', exact: true, name: 'Home', component: HomeContainer });
		app.Router.addRoute({ path: '/Table', exact: true, name: 'Table', component: TableContainer });
		app.Navigation.addItem({
			name: 'Home',
			url: '/',
			icon: 'cil-home',
		})
		app.Navigation.addItem({
			name: 'Table',
			url: '/Table',
			icon: 'cil-chart',
		})
	}
}
```

Pe5h4 commented on doc/datatable.md:
```diff
@@ -1,27 +1,36 @@
 
+DataTable can be mounted in its own container as it is implemented in demo app or as a widget in another app's container. In first case you must create container for `DataTable` and connect it to Home Module of your application by adding it's route and navigation item. Example may be found in `demo/src/modules/home/index.js`.
+
 DataTable needs obligatory props `headers`, `data`, `count`, `currentPage`, `setPage` and `title` to render.
```
@titkov49 I would also add the example here:

```
   ...
	return (
		<Container>
			<DataTable
				title={{text: "Table Demo", icon: 'cil-user'}}
				data={data}
				headers={headers}
				count={count}
				currentPage={page}
				setPage={setPage}
			/>
		</Container>
	)
   ...
```

Pe5h4 commented on doc/datatable.md:
```diff
@@ -1,27 +1,36 @@
+
 DataTable needs obligatory props `headers`, `data`, `count`, `currentPage`, `setPage` and `title` to render.
 
+Some props can be obtained from configuration of the project  (e.g. in demo app `headers` and `limit` props are obtained from the configuration).
```
@titkov49 Here, an example would be cool:

```
let ConfigDefaults = {
    ...
	table: {
		headers: [ { name: 'Name', key: 'username' }, { name: 'Provider', key: '_provider_id' }, { name: 'Type', key: '_type' } ],
		limit: 10
	},
    ...
};
```

Pe5h4 commented on doc/datatable.md:
```diff
@@ -1,27 +1,36 @@
 
+Some props can be obtained from configuration of the project  (e.g. in demo app `headers` and `limit` props are obtained from the configuration).
+
 Prop `title` is an object that has obligatory property `text` which is string and it will be rendered as table title. Also `title` has optional property `icon` for rendering icon alongside the table title and which can be either React Component (e.g. from package `@material-ui/icons`) or it can also be a string which is classname for icon element and will be rendered as <i className={icon}></i>
```
@titkov49 Add an example:

```
title={{text: "Table Demo", icon: 'cil-user'}}
```

Pe5h4 commented on doc/datatable.md:
```diff
@@ -1,27 +1,36 @@
 
-Prop `headers` is basically an array containing objects with obligatory properties `text` and `type`. Property `text` is a string that will be rendered as a header (first letter of `headers.text` automaticly will be converted to uppercase) in headers row of the table. Property `type` is a type of data in each cell for column of that header, it can be either `string` or `json`.
-Order of headers in a table is the same as in prop `headers`.
+Prop `headers` is basically an array containing objects with obligatory properties `name` and `key`. Property `name` is a string that will be rendered as a header cell in headers row of the table. Property `key` is a key name for getting data from `data` prop, it must be the same as it is in objects in `data` prop.
```
@titkov49 Add an example:

```
headers: [ { name: 'Name', key: 'username' }, { name: 'Provider', key: '_provider_id' }, { name: 'Type', key: '_type' } ],
```

Pe5h4 commented on doc/datatable.md:
```diff
@@ -1,27 +1,36 @@
 
-Prop `data` is an array of objects. `DataTable` goes through `data` and looks for those properties which has the same name as 
-`headers[idx].text`, if such exist, it will be rendered under that `headers[idx].text` as string or json.
+Prop `data` is an array of objects. `DataTable` component goes through objects in `data` and for each object looks for those properties which has the same key name that was defined in `headers[idx]`. If properties exist in that object from `data`, they will be rendered as string under the appropriate headers. Otherwise if some properties doesn't exist in that object, the `-` symbol will be render ...
```
@titkov49 I would write here, that data are usually obtained from API endpoint with the use of axios and add there a simple example, e.g:

```
   ...
	const retrieveData = () => {
		Axios.get()
		.then(response => {
			setData(response.data);
		})
		.catch(e => {
			console.log(e);
			props.app.addAlert("warning", "Failed to fetch the data");
		});
	};
   ...
```

Pe5h4 commented on doc/datatable.md:
```diff
@@ -1,27 +1,36 @@
+Prop's `data` length should be less or equal to 10 or to `limit` prop if such is defined. If `data` length is more than `limit`, then it will be cutted. 
+Array `data` may be fetched from `axios` in container of the `DataTable` for example. In demo we used predefined `initData` object contains data collection and `fetch(limit, page, str)` method that accepted Table's limit, current page and string from search input field. The `fetch` method is immitating fetch request from the server. You may take a look at our demo app to see how data for `data` p ...
 
 Also `DataTable` needs props `count`, `currentPage` and `setPage`. Prop `count` is number representing the count of all items, it is used to calculate last page in pagination of `DataTable`. Props `currentPage` and `setPage` are number and function respectively, first represents current page of `DataTable`, second is used to navigate between pages in `DataTable`.
```
@titkov49 I would also write here, that the amount of data for pagination can be obtained from API endpoint for every page when set as a parameter in the request, e.g.

```
Axios.get("", {params: {p:page, i: limit, f: search}})
```

where `f` parameter is the search or filter value for `search` prop

Pe5h4 commented on doc/datatable.md:
```diff
@@ -1,27 +1,36 @@
-Props `limit` and `setLimit` is used to handle limiting the amount of objects per page. Prop `limit` is number and represents the amount of rendered objects per page and calculate amount of pages in `DataTable`, default value is 20. Prop `setLimit` is a function for changing `limit`. If `setLimit` exists `DataTable` renders dropdown box for setting limit with values 5, 10, 15, 20.
+Props `limit` and `setLimit` is used to handle limiting the amount of objects per page. Prop `limit` is number and represents the amount of rendered objects per page and calculate amount of pages in `DataTable`, default value is 10. Prop `setLimit` is a function for changing `limit`. If `setLimit` exists `DataTable` renders dropdown box for setting limit with values 5, 10, 15, 20.
 
 Prop `createButton` is used for rendering button with a link for redirecting to create page. It is an object with obligatory properties `text` and `pathname` which are text inside the button and pathname for redirecting respectively.
```
@titkov49 Please add a settings example

Pe5h4 replied:
also for `search`

Pe5h4 commented on src/containers/DataTable/Pagination.js:
```diff
@@ -68,6 +68,6 @@ export default function ({ currentPage, setPage, lastPage }) {
-  );
+		</Pagination>
+	);
 }
```
@titkov49 Add a newline

Pe5h4 commented on src/containers/DataTable/Table.js:
```diff
@@ -1,51 +1,68 @@
+
+	return (
+		<Table size="sm">
+		<Headers headers={headersRow} />
```
@titkov49 Indent components as following:

```
		<Table size="sm">
		        <Headers headers={headersRow} />
		        <tbody>
			        {TableRows}
		        </tbody>
		</Table>
```

Pe5h4 commented on src/containers/DataTable/Table.js:
```diff
@@ -1,51 +1,68 @@
\ No newline at end of file
+const mapStateToProps = (state) => ({ advmode: state.advmode.enabled });
+
+export default connect(mapStateToProps)(ASABTable);
```
@titkov49 Add a newline

Pe5h4 commented on src/containers/DataTable/index.js:
```diff
@@ -3,130 +3,124 @@ import { Link } from 'react-router-dom';
+		</Col>
+		</Row>
+	);
+};
```
@titkov49 Add a newline

Pe5h4 reviewed (changes requested):
@titkov49 Just a request for small changes, mostly in the documentation. Otherwise it is good

titkov49 commented:
@Pe5h4 @ateska @wuujcik 
Changes: 
- Added new lines at the end of the files
- Added more examples to docs
- Delete unused **mapStateToProps** from **TableContainer.js**

Pe5h4 commented on doc/datatable.md:
````diff
@@ -0,0 +1,350 @@
+};
+```
+
+Props `search` and `onSearch` is used to render input search box in `DataTable`. First prop `search` is either a boolean or an object containing optional property `placeholder` for placeholder of search input box and optional property `icon` which can either be a React Component (e.g. from package `@material-ui/icons`) or a string which is className and will be rendered as <i className={icon}></i ...
````
@titkov49 onSearch will search only among the items on the current page, am I right? It might be good if you can at least mention here, that the search can be done also on axios request to the database. (with example)
